src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c

```diff
@@ -0,0 +1,393 @@
+ * SUNDIALS Copyright Start
+ * Copyright (c) 2002-2023, Lawrence Livermore National Security
+ * and Southern Methodist University.
+ * All rights reserved.
+ *
+ * See the top-level LICENSE and NOTICE files for details.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ * SUNDIALS Copyright End
+ * -----------------------------------------------------------------
+ * This is the implementation file for the
+ * SUNAdaptController_Soderlind module.
+ * -----------------------------------------------------------------*/
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <sunadaptcontroller/sunadaptcontroller_soderlind.h>
+#include <sundials/sundials_math.h>
+
+
+/* ---------------
+ * Macro accessors
+ * --------------- */
+
+#define SODERLIND_CONTENT(C)     ( (SUNAdaptControllerContent_Soderlind)(C->content) )
+#define SODERLIND_K1(C)          ( SODERLIND_CONTENT(C)->k1 )
+#define SODERLIND_K2(C)          ( SODERLIND_CONTENT(C)->k2 )
+#define SODERLIND_K3(C)          ( SODERLIND_CONTENT(C)->k3 )
+#define SODERLIND_K4(C)          ( SODERLIND_CONTENT(C)->k4 )
+#define SODERLIND_K5(C)          ( SODERLIND_CONTENT(C)->k5 )
+#define SODERLIND_BIAS(C)        ( SODERLIND_CONTENT(C)->bias )
+#define SODERLIND_EP(C)          ( SODERLIND_CONTENT(C)->ep )
+#define SODERLIND_EPP(C)         ( SODERLIND_CONTENT(C)->epp )
+#define SODERLIND_HP(C)          ( SODERLIND_CONTENT(C)->hp )
+#define SODERLIND_HPP(C)         ( SODERLIND_CONTENT(C)->hpp )
+#define SODERLIND_FIRSTSTEPS(C)  ( SODERLIND_CONTENT(C)->firststeps )
+
+/* ------------------
+ * Default parameters
+ * ------------------ */
+
+#define DEFAULT_K1          RCONST(1.25)    /* H_{0}321 parameters */
+#define DEFAULT_K2          RCONST(0.5)
+#define DEFAULT_K3          RCONST(-0.75)
+#define DEFAULT_K4          RCONST(0.25)
+#define DEFAULT_K5          RCONST(0.75)
+#define DEFAULT_PID_K1      RCONST(0.58)    /* PID parameters */
+#define DEFAULT_PID_K2     -RCONST(0.21)
+#define DEFAULT_PID_K3      RCONST(0.1)
+#define DEFAULT_PI_K1       RCONST(0.8)     /* PI parameters */
+#define DEFAULT_PI_K2      -RCONST(0.31)
+#define DEFAULT_I_K1        RCONST(1.0)     /* I parameters */
+#define DEFAULT_EXPGUS_K1   RCONST(0.367)   /* Explicit Gustafsson parameters */
+#define DEFAULT_EXPGUS_K2   RCONST(0.268)
+#define DEFAULT_IMPGUS_K1   RCONST(0.98)    /* Implicit Gustafsson parameters */
+#define DEFAULT_IMPGUS_K2   RCONST(0.95)
+#define DEFAULT_BIAS        RCONST(1.5)
+#define TINY                RCONST(1.0e-10)
+
+
+/* -----------------------------------------------------------------
+ * exported functions
+ * ----------------------------------------------------------------- */
+
+/* -----------------------------------------------------------------
+ * Function to create a new Soderlind controller
+ */
+
+SUNAdaptController SUNAdaptController_Soderlind(SUNContext sunctx)
+{
+  SUNAdaptController C;
+  SUNAdaptControllerContent_Soderlind content;
```

Comment:
```suggestion
  SUNAdaptControllerContent_Soderlind content;

  if (sunctx == NULL) { return NULL; }
```

Reply:
It is redundant but I don't think we should rely on a call to another function for input checks.